private/tools/java/rules/jvm/external/jar/AddJarManifestEntry.java

```diff
@@ -0,0 +1,171 @@
+         ZipInputStream zis = new ZipInputStream(fis)) {
+
+      try (OutputStream fos = Files.newOutputStream(out);
+           ZipOutputStream zos = new JarOutputStream(fos)) {
+
+        ZipEntry sourceEntry;
+        while ((sourceEntry = zis.getNextEntry()) != null) {
+          String name = sourceEntry.getName();
+
+          ZipEntry outEntry = new ZipEntry(name);
+          outEntry.setMethod(sourceEntry.getMethod());
+          outEntry.setTime(sourceEntry.getTime());
+          outEntry.setComment(sourceEntry.getComment());
+          outEntry.setExtra(sourceEntry.getExtra());
+
+          if (manifestEntry != null && JarFile.MANIFEST_NAME.equals(name)) {
+            Manifest manifest = new Manifest(zis);
+            String[] manifestEntryParts = manifestEntry.split(":", 2);
+            manifest.getMainAttributes().put(new Attributes.Name(manifestEntryParts[0]), manifestEntryParts[1]);
+
+            if (sourceEntry.getMethod() == ZipEntry.STORED) {
+              CRC32OutputStream crc32OutputStream = new CRC32OutputStream();
+              manifest.write(crc32OutputStream);
+              outEntry.setSize(crc32OutputStream.getSize());
+              outEntry.setCrc(crc32OutputStream.getCRC());
+            }
+            zos.putNextEntry(outEntry);
+            manifest.write(zos);
+            manifestUpdated = true;
+
+          } else {
+
+            if (sourceEntry.getMethod() == ZipEntry.STORED) {
+              outEntry.setSize(sourceEntry.getSize());
+              outEntry.setCrc(sourceEntry.getCrc());
+            }
+
+            try {
+              zos.putNextEntry(outEntry);
+            } catch (ZipException e) {
+              if (e.getMessage().contains("duplicate entry:")) {
+                // If there is a duplicate entry we keep the first one we saw.
+                verboseLog("WARN: Skipping duplicate jar entry " + outEntry.getName() + " in " + source);
+                continue;
+              } else {
+                throw e;
+              }
+            }
+            while ((bytesRead = zis.read(buffer)) != -1) {
+              zos.write(buffer, 0, bytesRead);
+            }
+          }
+        }
+
+        if (manifestEntry != null && !manifestUpdated) {
+          verboseLog("INFO: No jar manifest found in " + source + " Adding new jar manifest.");
+          Manifest manifest = new Manifest();
+          manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0");
+          manifest.getMainAttributes().put(new Attributes.Name("Created-By"), "AddJarManifestEntry");
+          String[] manifestEntryParts = manifestEntry.split(":", 2);
+          manifest.getMainAttributes().put(new Attributes.Name(manifestEntryParts[0]), manifestEntryParts[1]);
+
+          ZipEntry newManifestEntry = new ZipEntry(JarFile.MANIFEST_NAME);
+          newManifestEntry.setTime(DEFAULT_TIMESTAMP);
+          zos.putNextEntry(newManifestEntry);
+          manifest.write(zos);
```

Comment:
`JarInputStream` will only look at the first two entries of a zip when looking for a manifest. You need to write the manifest first, and _then_ write the contents of the jar.

Reply:
This is a good use case but I don't think `JarInputStream` is used by anything in Bazel.  Is the goal to support external tools that might use `JarInputStream` to work with stamped jars?

Reply:
Updated so if the manifest is missing it will be added as the first entry in the new jar.